### PR TITLE
Download support

### LINKED ISSUE #222: 'download' button
we should either remove it for now or get it to work.

### PR DESCRIPTION
**Pull request by clkao**
Closes #222

### CODE REVIEW

File: client/js/models/archive.js [hunk 1]
```diff
@@ -235,6 +237,24 @@ module.exports = {
       var archive = state.instance
       var readStream = archive.createFileReadStream(data.entryName)
       done(readStream)
+    },
+    download: function (data, state, send, done) {
+      const archive = state.instance
+      const zip = new Jszip()
+      zip.file('Hello.txt', 'Hello World\n')
```
Comment (karissa) [inline client/js/models/archive.js]: this line is probably an accident ya?

File: client/js/models/archive.js [hunk 2]
```diff
@@ -235,6 +237,24 @@ module.exports = {
...
+      Object.keys(state.entries).sort().forEach((key) => {
+        const entry = state.entries[key]
+        if (entry.type === 'directory') {
+          // XXX: might need to manually create things
```
Comment (karissa) [inline client/js/models/archive.js]: ideally these entries have some sort of reference to their full path in their object, if they don't already we should make that happen

File: client/js/models/archive.js (hunk 1 above)
Comment (clkao) [inline client/js/models/archive.js]: lol ya

File: client/js/models/archive.js (hunk 2 above)
Comment (clkao) [inline client/js/models/archive.js]: i thought key is the full path inside archive?

**Comment by laurengarcia** [comment]
I didn't get a chance to pull this down today but it looks awesome, can't wait to check it out.

**Comment by karissa** [comment]
lgtm!

File: client/js/models/archive.js [hunk 3]
```diff
@@ -235,6 +237,23 @@ module.exports = {
       var archive = state.instance
       var readStream = archive.createFileReadStream(data.entryName)
       done(readStream)
+    },
+    download: function (data, state, send, done) {
+      const archive = state.instance
+      const zip = new Jszip()
+      Object.keys(state.entries).sort().forEach((key) => {
```
Comment (karissa) [inline client/js/models/archive.js]: do we have to sort them?

Comment (clkao) [inline client/js/models/archive.js]: not really, but it'll be easier to test things.

**Comment by karissa** [comment]
I say we merge this under the condition that we have a message that it can only download archives under 500MB right now. If the archive is over 500MB, we should tell the user to use the cli to download the files.

I started a branch here to enable downloads of up to 15GB, but there are bugs with the polyfill package: https://github.com/datproject/dat.land/tree/download-streamsaver

**Pull request merged**